ai_server/server.py

```diff
@@ -1,12 +1,250 @@
+            cmd,
+            capture_output=True,
+            text=False,
+            timeout=timeout,
+            check=True
+        )
+        
+        # For V3 IQ1_M and IQ2_XXS models
+        try:
+            stdout_text = result.stdout.decode('utf-8')
+        except UnicodeDecodeError:
+            stdout_text = result.stdout.decode('utf-8', errors='replace')
+
+        output_lines = stdout_text.strip().split('\n')
+        response_lines = []
+        for line in output_lines:
+            if any(skip_pattern in line for skip_pattern in LLAMA_CPP_SKIP_PATTERNS):
+                continue
+            
+            if line.startswith('llama_perf_'):
+                break
+                
+            if line.strip():
+                response_lines.append(line.strip())
+        
+        response = ' '.join(response_lines) if response_lines else ""
+        return response if response else "No response generated."
+        
+    except subprocess.TimeoutExpired:
+        raise Exception(f"Llama.cpp request timed out for model {model}")
+    except subprocess.CalledProcessError as e:
+        stderr_text = ""
+        if e.stderr:
+            try:
+                stderr_text = e.stderr.decode('utf-8')
+            except UnicodeDecodeError:
+                stderr_text = e.stderr.decode('utf-8', errors='replace')
+        raise Exception(f"Llama.cpp failed for {model}: {stderr_text.strip() if stderr_text else 'Unknown error'}")
+    except FileNotFoundError:
+        raise Exception("Llama.cpp CLI not found")
+
+def chat_with_model(model: str, content: str, llama_mode: str = "cli") -> str:
+    """Route chat request based on llama_mode: server (external), cli, or default fallback."""
+    if llama_mode == "server":
+        # Use external llama-server with fallbacks
+        if not LLAMA_SERVER_URL:
+            raise Exception("LLAMA_SERVER_URL environment variable not set for server mode")
+        
+        if is_llamacpp_available(model):
+            try:
+                return chat_with_llama_server_http(model, content)
+            except Exception:
+                # Fallback to llama-cli if server fails
+                try:
+                    return chat_with_llamacpp(model, content)
+                except Exception:
+                    # Final fallback to ollama
+                    return chat_with_ollama(model, content)
+        else:
+            # Model not available in llama.cpp, use ollama
+            return chat_with_ollama(model, content)
+    
+    elif llama_mode == "cli":
+        if is_llamacpp_available(model):
+            try:
+                return chat_with_llamacpp(model, content)
+            except Exception:
+                # Fallback to ollama if llama-cli fails
+                try:
+                    return chat_with_ollama(model, content)
+                except Exception as e_ollama:
+                    available_llamacpp = get_available_llamacpp_models()
```

Comment:
It's not a bad idea to provide users with a list of available models, but I don't think it should be in an error message.

Please remove this function from this PR; in this error handling you can just provide the error message from the exception.

In a _separate PR_ (after you complete this on), you can restore the function but call it in a separate top-level route called `available_models`, and return in JSON format.